Improve structure of specification

danielfett opened the pull request:
To fix Issue #290, this PR reorganizes what was previously section 5 as follows:

![Screenshot_20231004_144710](https://github.com/oauth-wg/oauth-selective-disclosure-jwt/assets/89476/0c3fab16-e957-4ff3-acc0-3cd0691903e4)

danielfett commented:
Note: I also changed the story in Section 5 from a bottom-up to a top-down approach (structure of compact serialization SD-JWT first, then going into the details).

bc-pi commented on draft-ietf-oauth-selective-disclosure-jwt.md:
````diff
@@ -242,11 +242,42 @@ At a high level, the Verifier
+The serialized format for the SD-JWT is the concatenation of each part delineated with a single tilde ('~') character as follows:
+
+```
+<Issuer-signed JWT>~<Disclosure 1>~<Disclosure 2>~...~<Disclosure N>~<optional KB-JWT>
````
```suggestion
<Issuer-signed JWT>~<Disclosure 1>~<Disclosure 2>~...~<Disclosure N>~<optional KB-JWT>

```

adding a newline so (hopefully) the horizontal scrollbar in the HTML output doesn't cover up the content

bc-pi commented on draft-ietf-oauth-selective-disclosure-jwt.md:
````diff
@@ -422,9 +453,47 @@ disclosable array elements for which they did not receive a Disclosure. In the
+
+To validate the signature on the Key Binding JWT, the Verifier MUST use the key material in the SD-JWT. If it is not clear from the SD-JWT, the Key Binding JWT MUST specify which key material the Verifier needs to use to validate the Key Binding JWT signature using JOSE header parameters such as `kid` and `x5c`.
+
+Below is a non-normative example of a Key Binding JWT header:
+
+```
+{
+  "alg": "ES256",
+  "typ": "kb+jwt"
+}
+```
+
````
With the other examples gone from this section - this one feels a bit out of place. And not needed. 
```suggestion
```

Sakurann replied:
should this example be re-added to another section? it was added because people wanted to see an example of a KB-JWT header..

danielfett replied:
I agree that removing makes most sense. We're not showing any other headers in the document, there is nothing surprising in this header, and people can still decode the KB-JWT if they like to see a header.

bc-pi reviewed (commented):
This looks pretty good and does generally improve the structure. 
The Decoy Digests, Hash Function Claim, Holder Public Key Claim subsections feel a bit out of place in section 7.  What would you think about moving them to subsections under 5.1.  Issuer-signed JWT Payload? The ToC might look something like this:

![Screenshot 2023-10-04 at 3 08 50 PM](https://github.com/oauth-wg/oauth-selective-disclosure-jwt/assets/71398439/2990ded2-aba3-4559-bd28-e7b902a63477)

bc-pi reviewed (changes requested):
I'd like to see a few more subsections moved https://github.com/oauth-wg/oauth-selective-disclosure-jwt/pull/347#pullrequestreview-1658506950 and one minor example removed https://github.com/oauth-wg/oauth-selective-disclosure-jwt/pull/347/files#r1346423276 but otherwise this looks good.

Sakurann commented on draft-ietf-oauth-selective-disclosure-jwt.md:
```diff
@@ -242,11 +242,43 @@ At a high level, the Verifier
+In the case that there is no Key Binding JWT, the last element MUST be an empty string and the last separating tilde character MUST NOT be omitted.
 
-## SD-JWT Payload
+The Disclosures are linked to the SD-JWT payload through the
+digest values included therein.
```
```suggestion
The Disclosures are linked to the Issuer-signed JWT by including
Disclosure digest values therein.
```

Sakurann commented on draft-ietf-oauth-selective-disclosure-jwt.md:
```diff
@@ -242,11 +242,43 @@ At a high level, the Verifier
+The Disclosures are linked to the SD-JWT payload through the
+digest values included therein.
+
+When issued to a Holder, the Issuer includes all the relevant Disclosures in the SD-JWT.
+
+For presentation to a Verifier, the Holder sends the SD-JWT including only its selected
+set of the Disclosures to the Verifier.
```
```suggestion
When issuing to a Holder, the Issuer includes all the relevant Disclosures in the SD-JWT.

When presenting to a Verifier, the Holder sends only the selected set of the Disclosures in the SD-JWT.
```
consistency here felt better?

Sakurann commented on draft-ietf-oauth-selective-disclosure-jwt.md:
```diff
@@ -242,11 +242,43 @@ At a high level, the Verifier
+set of the Disclosures to the Verifier.
+
+The Holder MAY send any subset of the Disclosures to the Verifier, i.e.,
+none, multiple, or all Disclosures. For data that the Holder does not want to reveal
```
```suggestion
none, parts of, or all Disclosures. For data that the Holder does not want to reveal
```
`parts of ` feels a little off, but `all` is part of `multiple`..

bc-pi replied:
`some of` ?

danielfett replied:
```suggestion
none, some, or all Disclosures. For data that the Holder does not want to reveal
```
WDYT?

bc-pi replied:
WFM

Sakurann commented on draft-ietf-oauth-selective-disclosure-jwt.md:
```diff
@@ -363,14 +395,14 @@ The SHA-256 digest of the Disclosure
-## Embedding Disclosure Digests in SD-JWTs {#embedding_disclosure_digests}
+### Embedding Disclosure Digests in SD-JWTs {#embedding_disclosure_digests}
 
 For selectively disclosable claims, the digests of the Disclosures are embedded into the SD-JWT instead of the claims themselves. The precise way of embedding depends on whether a claim is an object property (key-value pair) or an array element.
```
```suggestion
For selectively disclosable claims, the digests of the Disclosures are embedded into the Issuer-signed JWT instead of the claims themselves. The precise way of embedding depends on whether a claim is an object property (key-value pair) or an array element.
```

Sakurann commented on draft-ietf-oauth-selective-disclosure-jwt.md:
```diff
@@ -451,25 +521,25 @@ The payload is then signed by the Issuer to create a JWT like the following:
+
+<{{examples/simple/kb_jwt_payload.json}}
+
+# Data Format Details
```
it almost feels like `# Advanced Features`

Sakurann commented:
> The Decoy Digests, Hash Function Claim, Holder Public Key Claim subsections feel a bit out of place in section 7. What would you think about moving them to subsections under 5.1. Issuer-signed JWT Payload?

+1 to this

Sakurann reviewed (changes requested):
reads so much better and flows more naturally - agree with Brian's suggested subsections 5.1.1~5.1.3. also made few suggestions

danielfett commented:
I made the additional changes proposed by Brian, with the following deltas:

* I moved the section "Decoy Digests" to be the last subsection of "5.2 Disclosures". It did not make sense in 5.1 as not all concepts were introduced there yet.
* I slightly reformatted the list of requirements on the SD-JWT payload in Section 5.1.
* I noticed that the note on line breaks in the examples is not in the correct place and moved it towards the beginning of Section 6 (Example 1).

bc-pi commented:
> I moved the section "Decoy Digests" to be the last subsection of "5.2 Disclosures". It did not make sense in 5.1 as not all concepts were introduced there yet. 

That makes sense and is better than what I'd suggested.
